src/main.rs

```diff
@@ -1665,16 +1634,98 @@ fn foreign_packages<'a>(
     metadata
         .packages
         .iter()
-        .filter(|package| package.is_third_party(&config.audit_as_crates_io))
+        .filter(|package| package.is_third_party(&config.policy))
 }
 
-fn first_party_packages<'a>(
+/// All first-party packages, **without** the audit-as-crates-io policy applied
+/// (because it's used for validating that field's value).
+fn first_party_packages_strict<'a>(
     metadata: &'a Metadata,
     config: &'a ConfigFile,
 ) -> impl Iterator<Item = &'a Package> + 'a {
     // Only analyze things from crates.io (no source = path-dep / workspace-member)
     metadata
         .packages
         .iter()
-        .filter(|package| !package.is_third_party(&config.audit_as_crates_io))
+        .filter(|package| !package.is_third_party(&config.policy))
+}
+
+fn check_audit_as_crates_io(
+    out: &mut dyn Write,
+    cfg: &Config,
+    store: &Store,
+) -> Result<(), VetError> {
+    let cache = Cache::acquire(cfg)?;
+    let mut needs_audit_as_entry = vec![];
+    let mut shouldnt_be_audit_as = vec![];
+
+    'packages: for package in first_party_packages_strict(&cfg.metadata, &store.config) {
+        let audit_policy = store
+            .config
+            .policy
+            .get(&package.name)
+            .and_then(|policy| policy.audit_as_crates_io);
+        if audit_policy == Some(false) {
+            // They've explicitly said this is first-party so we don't care about what's in the registry
+            continue;
+        }
+
+        if let Some(index_entry) = cache.query_package_from_index(&package.name) {
+            for index_version in index_entry.versions() {
+                if let Ok(index_ver) = index_version.version().parse::<cargo_metadata::Version>() {
+                    if index_ver == package.version {
```

Comment:
FWIW I think that exposing a version of `query_package_from_index` which takes a name and a version might be nice (`query_package_version_from_index`?). When doing the checksum validation stuff, I had to add something like that as well, because I just wanted a specific version of the crate, and it's a bit fiddly to write out.

Reply:
I added this, but it has to be a two-pass thing since crates-index returns a wholy owned value for a Crate and subsequent Version iteration borrows from that value.

Reply:
You can clone the `Version` value and return it as an owned value. They shouldn't be that large or expensive to clone out and the UX improvement seems pretty nice.